x-pack/plugins/security_solution/server/endpoint/routes/resolver/utils/fetch.ts

```diff
@@ -102,29 +139,72 @@ export class Fetcher {
     return tree;
   }
 
+  private async getNode(entityID: string): Promise<LifecycleNode | undefined> {
+    const query = new LifecycleQuery(this.indexPattern, this.endpointID);
+    const results = await query.search(this.client, entityID);
+    if (results.length === 0) {
+      return;
+    }
+
+    return createLifecycle(entityID, results);
+  }
+
+  private static getAncestryAsArray(event: ResolverEvent): string[] {
+    const ancestors = ancestryArray(event);
+    if (ancestors) {
+      return ancestors;
+    }
+
+    const parentID = parentEntityId(event);
+    if (parentID) {
+      return [parentID];
+    }
+
+    return [];
+  }
+
   private async doAncestors(
-    curNodeID: string,
+    ancestors: string[],
     levels: number,
     ancestorInfo: ResolverAncestry
   ): Promise<void> {
-    if (levels === 0) {
-      ancestorInfo.nextAncestor = curNodeID;
+    if (levels <= 0) {
       return;
     }
 
     const query = new LifecycleQuery(this.indexPattern, this.endpointID);
-    const results = await query.search(this.client, curNodeID);
+    const results = await query.search(this.client, ancestors);
 
     if (results.length === 0) {
+      ancestorInfo.nextAncestor = null;
       return;
     }
-    ancestorInfo.ancestors.push(createLifecycle(curNodeID, results));
 
-    const next = parentEntityId(results[0]);
-    if (next === undefined) {
-      return;
-    }
-    await this.doAncestors(next, levels - 1, ancestorInfo);
+    // bucket the start and end events together for a single node
+    const ancestryNodes = results.reduce(
+      (nodes: Map<string, LifecycleNode>, ancestorEvent: ResolverEvent) => {
+        const nodeId = entityId(ancestorEvent);
+        let node = nodes.get(nodeId);
+        if (!node) {
+          node = createLifecycle(nodeId, []);
+        }
+
+        node.lifecycle.push(ancestorEvent);
+        return nodes.set(nodeId, node);
+      },
+      new Map()
+    );
+
+    // the order of this array is going to be weird, it will look like this
+    // [furthest grandparent...closer grandparent, next recursive call furthest grandparent...closer grandparent]
+    ancestorInfo.ancestors.push(...ancestryNodes.values());
+    ancestorInfo.nextAncestor = parentEntityId(results[0]) || null;
+    const levelsLeft = levels - ancestryNodes.size;
```

Comment:
This should never lead to the levels being less than 0 right?

Reply:
It would likely be a bug, but if `levels` was passed in as say 2 and we passed in a `ancestors` array with more than 2 values, elasticsearch would respond with the requested number and `levels - ancestryNodes.size` would result in a negative number.

So I figured it'd probably be best to be extra careful :)

Reply:
makes sense :)